src/Compilers/VisualBasic/Portable/Symbols/ConstraintsHelper.vb

```diff
@@ -895,28 +894,34 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
                     End If
 
                 Case Else
-                    If typeArgument.TypeKind = TypeKind.Class Then
+                    Dim isStructure As Boolean = typeArgument.TypeKind = TypeKind.Structure
+
+                    Dim constraintError = ConstructorConstraintError.NoPublicParameterlessConstructor
+
+                    If typeArgument.TypeKind = TypeKind.Class OrElse isStructure Then
                         Dim classType = DirectCast(typeArgument, NamedTypeSymbol)
+                        constraintError = HasPublicParameterlessConstructor(classType, synthesizedIfMissing:=isStructure)
 
-                        If HasPublicParameterlessConstructor(classType) Then
-                            If classType.IsMustInherit Then
-                                If diagnosticsBuilder IsNot Nothing Then
-                                    ' "Type argument '{0}' is declared 'MustInherit' and does not satisfy the 'New' constraint for type parameter '{1}'."
-                                    diagnosticsBuilder.Add(New TypeParameterDiagnosticInfo(typeParameter, ErrorFactory.ErrorInfo(ERRID.ERR_MustInheritForNewConstraint2, typeArgument, typeParameter)))
-                                End If
-                                Return False
-                            Else
-                                Return True
-                            End If
+                        If constraintError = ConstructorConstraintError.None AndAlso classType.IsMustInherit Then
+                            constraintError = ConstructorConstraintError.MustInheritType
                         End If
-
                     End If
 
                     If diagnosticsBuilder IsNot Nothing Then
-                        ' "Type argument '{0}' must have a public parameterless instance constructor to satisfy the 'New' constraint for type parameter '{1}'."
-                        diagnosticsBuilder.Add(New TypeParameterDiagnosticInfo(typeParameter, ErrorFactory.ErrorInfo(ERRID.ERR_NoSuitableNewForNewConstraint2, typeArgument, typeParameter)))
+                        Select Case constraintError
```

Comment:
>Select Case constraintError [](http://example.com/codeflow?start=24&length=27)

Consider adding case for None and a throw for an unexpected value in Case Else